### PR TITLE
Issue 195 Acceptance test for service-manager using vagrant-spec

### PR DESCRIPTION
**Pull request by hferentschik**
Here is a first cut of the acceptance test harness. I want to add some more tests and maybe make some other adjustments, but I also want to share it early to get some feedback and making sure that everyone can run it.

The tests can be run against VirtualBox or Libvirt (the latter with a caveat). If you have the box file you want to use for testing already on your system, you can run the tests like:

```
$ export VAGRANT_SPEC_BOX=file://<full path to box file>
# VirtualBox
$ bundle exec rake acceptance
# Libvirt
$ bundle exec rake acceptance['libvirt']
```

Obviously the box file used needs to be the correct one for the provider. I added another helper task which can download the publicly available box files (might be nice for CI testing!?)

```
$ export REDHAT_USER=<any user who can login into access.redhat.com>
$ export REDHAT_PASSWORD=<mypass>
$ bundle exec rake getbox
# or
$ bundle exec rake getbox['libvirt']
```

Regarding the mentioned caveat. The current tests pass on OS X using the VirtualBox box. When I run the same test on Linux using Libvirt, I get one test failure. The gist seems to be that when using VirtualBox I get _DOCKER_API_VERSION=1.21_ as part of the docker env output whereas with Libvirt I get _DOCKER_API_VERSION=1.20_. 

WDYT?

### CODE REVIEW

**Comment by brgnepal** [comment]
@hferentschik Could you squash into one commit?

**Comment by hferentschik** [comment]
> Could you squash into one commit?

Why? There is a reason for the current commit structure. I am happy to move the Gemfile.lock into the main commit, provided we agree that this is the right thing to do. 

The separate README cleanup commit should stay separate though imo.

**Comment by brgnepal** [comment]
@hferentschik It would be easy to track in future by one commit what all files were changed rather than looking for individual ones. Its like name-spacing all your changes related to issue fix under PR subject commit message "Fix #195 Acceptance test for service-manager using vagrant-spec".

May be @LalatenduMohanty could add more lights on it.

**Comment by hferentschik** [comment]
> It would be easy to track in future by one commit what all files were changed rather than looking for individual ones.

True. There is a middle ground, but usually I try to separate "cleanup" and refactoring commits from the commits which change functionality. Squashing all commits for a given issue into a single commit is rarely a good idea. It also makes reviewing harder. That said, there is somewhere a middle ground. Micro commits are useless as well. 

Also, as I mentioned, this is still work in progress and while it is I am not going to squash anything. Squashing is easy, splitting again out into multiple commits is much harder. So until we agree on the final details of the change, I'll keep the separate commits. Whether in the end it will get applied as single commit or multiple we can decide then

>  Its like name-spacing all your changes related to issue fix under PR subject commit message

This is exactly what I am doing. All my commits related to this issue have the "Issue #<key>" prefix.
In the Hibernate projects we had a policy to separate typo fixes, refactoring and so on from the actual issue fix. And even the issue fix itself could be multiple commits depending on the size of the change. Important is just that each commit still is "runnable" and passes the tests (think git bisect).

BTW, instead of whether to squash or not to squash we should talk about merge commits vs linear history. You want to improve your git history, I think we should adopt a linear, fast forward history only.

File: Rakefile [hunk 1]
```diff
@@ -1 +1,63 @@
...
+desc 'Download Vagrant box'
+task :getbox, [:provider] do |t, args|
+  if args[:provider].nil?
+    provider='virtualbox'
+  else
+    provider=args[:provider]
+  end
+  agent = Mechanize.new
+  agent.follow_meta_refresh = true
+  agent.get('https://access.redhat.com/downloads/content/293/ver=2/rhel---7/2.0.0/x86_64/product-software') do |page|
```
Comment (brgnepal) [inline Rakefile:18]: Would be better to extract it out in some constant like `RHEL_DOWNLOAD_LINK`

**Comment by hferentschik** [comment]
> Would be better to extract it out in some constant like `RHEL_DOWNLOAD_LINK`

+1 good idea

**Comment by LalatenduMohanty** [comment]
@hferentschik I think the  Gemfile.lock patch should be an independent patch. If you send the patch separately it is most likely to get merged relatively quicker than this patch too.

**Comment by LalatenduMohanty** [comment]
@hferentschik Have you given the patch a spin with ADB Vagrant box?

File: Rakefile (hunk 1 above)
Comment (LalatenduMohanty) [inline Rakefile:18]: And also adding a constant for  ADB download link . Not sure if we can use atlas. But we can use http://cloud.centos.org/centos/7/atomic/images/AtomicDeveloperBundle-1.8.0-CentOS7-VirtualBox.box

File: Gemfile [hunk 2]
```diff
@@ -3,13 +3,20 @@ source 'https://rubygems.org'
 gemspec
 
 group :development do
-  gem 'vagrant', git: 'https://github.com/mitchellh/vagrant.git'
+  gem 'vagrant',      git: 'https://github.com/mitchellh/vagrant.git'
+  gem 'vagrant-spec', git: 'https://github.com/mitchellh/vagrant-spec.git'
+  gem 'gem-compare'
+  gem 'rake'
+  gem 'vagrant-vbguest'
```
Comment (LalatenduMohanty) [inline Gemfile:10]: We don't need vagrant-vbguest here.

File: Gemfile [hunk 3]
```diff
@@ -3,13 +3,20 @@ source 'https://rubygems.org'
...
+  gem 'vagrant-libvirt'
```
Comment (LalatenduMohanty) [inline Gemfile:11]: This is also not compulsory for Virtualbox provider

File: Gemfile [hunk 4]
```diff
@@ -3,13 +3,20 @@ source 'https://rubygems.org'
...
+
   # added as the vagrant component wouldn't build without it
   gem 'json'
-  gem 'rake'
-  gem 'bundler', '~> 1.6'
 end
 
 group :plugins do
   gem 'vagrant-service-manager', path: '.'
+  gem 'vagrant-registration'
```
Comment (LalatenduMohanty) [inline Gemfile:21]: vagrant-registration is not required for ADB.

File: Rakefile (hunk 1 above)
Comment (hferentschik) [inline Rakefile:18]: > And also adding a constant for ADB download link 

+1 I forgot about ADB. I'll have to tasks. One for CDK and one for ADB.

**Comment by hferentschik** [comment]
> I think the Gemfile.lock patch should be an independent patch. If you send the patch separately it is most likely to get merged relatively quicker than this patch too.

+1 I think it is a good idea to track this separately in a different issue.

File: Gemfile (hunk 2 above)
Comment (hferentschik) [inline Gemfile:10]: true true,  cut&paste error from vagrant-registration. I'll remove it.

File: Gemfile (hunk 3 above)
Comment (hferentschik) [inline Gemfile:11]: > This is also not compulsory for Virtualbox provider

what is not? Note, these gems are in the development group, so there is no harm for the actual runtime dependencies.

File: Gemfile (hunk 4 above)
Comment (hferentschik) [inline Gemfile:21]: That's an interesting one, for the CDK one needs it, even though one might get away without registration altogether for the things we want to test. I can give it a go.

**Comment by hferentschik** [comment]
> Have you given the patch a spin with ADB Vagrant box?

Not yet. I need to see what's the best way to do that. Either two separate specs and potentially Vagrant skeletons and share some common test file or a single spec file as we have now.

**Comment by hferentschik** [comment]
Closing this for now. Trying something else as well...

**Pull request closed without merging**